docs/secrets.md

```diff
@@ -0,0 +1,357 @@
+
+### Use-case: Secret visible to one container in a pod
+<a name="use-case-two-containers"></a> 
+
+Consider a program that needs to handle HTTP requests, do some complex business
+logic, and then sign some messages with an HMAC.  Because it has complex
+application logic, there might be an unnoticed remote file reading exploit in
+the server, which could expose the private key to an attacker.
+
+This could be divided into two processes in two containers: a frontend container
+which handles user interaction and business logic, but which cannot see the
+private key; and a signer container that can see the private key, and responds
+to simple signing requests from the frontend (e.g. over localhost networking).
+
+With this partitioned approach, an attacker now has to trick the application
+server into doing something rather arbitrary, which may be harder than getting
+it to read a file.
+
+## Security Properties
+
+### Protections
+
+Because `secret` objects can be created independently of the `pods` that use
+them, there is less risk of the secret being exposed during the workflow of
+creating, viewing, and editing pods.  The system can also take additional
+precautions with `secret` objects, such as avoiding writing them to disk where
+possible. 
+
+A secret is only sent to a node if a pod on that node requires it.  It is not
+written to disk.  It is stored in a tmpfs.  It is deleted once the pod that
+depends on it is deleted.
+
+On most Kubernetes-project-maintained distributions, commuication between user
+to the apiserver, and from apiserver to the kubelets, is protected by SSL/TLS.
+Secrets are protected when transmitted over these channels.
+
+There may be secrets for several pods on the same node.  However, only the
+secrets that a pod requests are potentially visible within its containers.
+Therefore, one Pod does not have access to the secrets of another pod.  
+
+There may be several containers in a pod.  However, each container in a pod has
+to request the secret volume in its `volumeMounts` for it to be visible within
+the container.  This can be used to construct useful [security partitions at the
+Pod level](#use-case-two-containers).
+
+### Risks
+
+Applications still need to protect the value of secret after reading it from the volume,
+such not accidentally logging it or transmitting it to an untrusted party.
+
+A user who can create a pod that uses a secret can also see the value of that secret.  Even
+if apiserver policy does not allow that user to read the secret object, the user could
+run a pod which exposes the secret.
+If multiple replicas of etcd are run, then the secrets will be shared between them.
+By default, etcd does not secure peer-to-peer communication with SSL/TLS, though this can be configured.
+
+It is not possible currently to control which users of a kubernetes cluster can
+access a secret.  Support for this is planned. 
+**(Pmorie: is this currently controlled by namespace?)**
```

Comment:
Okay, figured out there aren't any yet.  That seems fine as we lack multiple users and policies right now anyways.